File: HR-js/graphSearch.js
class Node {
    constructor(id) {
        this.id = id;
        this.adjacent = [];
    }
}
class Graph { // directed graph
    constructor() {
        this.nodeLookup = new Map();
        this.nodeLookup.set(1, new Node(1));
        this.nodeLookup.set(2, new Node(2));
        this.nodeLookup.set(3, new Node(3));
        this.nodeLookup.set(4, new Node(4));
        this.nodeLookup.set(5, new Node(5));
        this.nodeLookup.set(6, new Node(6));
        this.addEdge(1, 2);
        this.addEdge(1, 3);
        this.addEdge(2, 4);
        this.addEdge(2, 5);
        this.addEdge(3, 4);
        this.addEdge(3, 5);
        this.addEdge(5, 6);
    }
    getNode(id) {
        return this.nodeLookup.get(id);
    }
    addEdge(source, destination) {
        let s = this.getNode(source);
        let d = this.getNode(destination);
        s.adjacent.push(d);
    }
    dfsHelper(source, destination) {
        let s = this.getNode(source);
        let d = this.getNode(destination);
        let visited = new Map();
        console.log(this.dfs(s, d, visited));
    }
    dfs(source, destination, visited) {
        if (visited.has(source.id)) {
            return false;
        }
        if (source === destination) {
            return true;
        }
        visited.set(source.id, true);
        console.log(visited);
        for (let i = 0; i < source.adjacent.length; i++) {
            if (this.dfs(source.adjacent[i], destination, visited)) {
                return true;
            }
        }
        return false;
    }
    bfsHelper(source, destination) {
        var s = this.getNode(source);
        var d = this.getNode(destination);
        console.log(this.bfs(s, d));
    }
    bfs(source, destination) {
        let visited = new Map();
        let nextToVisit = [];
        nextToVisit.push(source);
        while (nextToVisit.length > 0) {
            let node = nextToVisit.shift();
            if (node === destination) {
                return true;
            }
            if (visited.has(node.id)) {
                continue;
            }
            console.log(visited);
            visited.set(node.id, true);

            for (let i = 0; i < node.adjacent.length; i++) {
                nextToVisit.push(node.adjacent[i]);
            }
        }
        return false;
    }
}

var g = new Graph();
g.dfsHelper(1, 6);
// g.bfsHelper(1, 6);

File: HR-python/graphSearch.py
# Author: Shekhar Suman 18-05-18
# BFS and DFS implementation for graph
import queue 
class Node:
    def __init__(self, id):
        self.id = id
        self.adjacent = []

class Graph:
    def __init__(self):
        self.nodeLookup = dict()
        self.nodeLookup[1] = Node(1)
        self.nodeLookup[2] = Node(2)
        self.nodeLookup[3] = Node(3)
        self.nodeLookup[4] = Node(4)
        self.nodeLookup[5] = Node(5)
        self.nodeLookup[6] = Node(6)
        self.add_edge(1, 2)
        self.add_edge(1, 3)
        self.add_edge(2, 4)
        self.add_edge(2, 5)
        self.add_edge(3, 4)
        self.add_edge(3, 5)
        self.add_edge(5, 6)
    
    def get_node(self, id):
        return self.nodeLookup.get(id)

    def add_edge(self, source, destination):
        s = self.get_node(source)
        d = self.get_node(destination)
        s.adjacent.append(d)

    def dfs_helper(self, source, destination):
        s = self.get_node(source)
        d = self.get_node(destination)
        visited = dict()
        found = self.dfs(s, d, visited) 
        print(found)

    def dfs(self, source, destination, visited):
        if source.id in visited:
            return False

        if (source == destination):
            return True
        visited[source.id] = True
        print(visited)
        for child in source.adjacent:
            if self.dfs(child, destination, visited): # this returns the last recursive call value
                return True
        return False

    def bfs_helper(self, source, destination):
        s = self.get_node(source)
        d = self.get_node(destination)
        found = self.bfs(s, d) 
        print(found)

    def bfs(self, source, destination):
        visited = dict()
        q = queue.Queue()
        q.put(source)
        while not q.empty():
            node = q.get()
            if node == destination:
                return True
            if node.id in visited:
                continue
            print(visited)
            visited[node.id] = True
            for child in node.adjacent:
                q.put(child)
        return False



g = Graph()
# g.dfs_helper(1, 6)
g.bfs_helper(1, 5)